File: sdk/javascript/src/apis/projects.ts
/**
 * Projects API interface
 */

import type { AionClient } from '../client';
import type {
  Project,
  ProjectRequest,
  ProjectStatus,
  TemplateRequest,
  PaginatedResponse,
  PaginationParams,
  UUID,
} from '../types';

/**
 * Projects API interface for managing AION projects
 */
export class ProjectsAPI {
  constructor(private readonly client: AionClient) {}

  /**
   * List all projects with optional pagination
   */
  async list(params: PaginationParams = {}): Promise<PaginatedResponse<Project>> {
    const queryParams = {
      page: params.page ?? 1,
      per_page: params.perPage ?? 20,
      sort_order: params.sortOrder ?? 'asc',
      ...(params.sortBy && { sort_by: params.sortBy }),
    };

    const data = await this.client.get('/api/v1/projects', queryParams);
    return this.transformPaginatedResponse<Project>(data);
  }

  /**
   * Get a specific project by ID
   */
  async get(projectId: UUID): Promise<Project> {
    const data = await this.client.get(`/api/v1/projects/${projectId}`);
    return this.transformProject(data);
  }

  /**
   * Create a new project
   */
  async create(request: ProjectRequest): Promise<Project> {
    const data = await this.client.post('/api/v1/projects', {
      name: request.name,
      description: request.description,
      tech_stack: request.techStack,
      architecture: request.architecture,
      requirements: request.requirements,
      metadata: request.metadata,
    });
    return this.transformProject(data);
  }

  /**
   * Update an existing project
   */
  async update(projectId: UUID, request: ProjectRequest): Promise<Project> {
    const data = await this.client.put(`/api/v1/projects/${projectId}`, {
      name: request.name,
      description: request.description,
      tech_stack: request.techStack,
      architecture: request.architecture,
      requirements: request.requirements,
      metadata: request.metadata,
    });
    return this.transformProject(data);
  }

  /**
   * Delete a project
   */
  async delete(projectId: UUID): Promise<void> {
    await this.client.delete(`/api/v1/projects/${projectId}`);
  }

  /**
   * Generate a project from a template
   */
  async fromTemplate(request: TemplateRequest): Promise<Project> {
    const data = await this.client.post('/api/v1/projects/from-template', {
      template_id: request.templateId,
      project_name: request.projectName,
      customizations: request.customizations,
    });
    return this.transformProject(data);
  }

  /**
   * Get project execution status
   */
  async getStatus(projectId: UUID): Promise<ProjectStatus> {
    const data = await this.client.get(`/api/v1/projects/${projectId}/status`);
    return data as ProjectStatus;
  }

  /**
   * Start project execution
   */
  async start(projectId: UUID): Promise<void> {
    await this.client.post(`/api/v1/projects/${projectId}/start`, {});
  }

  /**
   * Stop project execution
   */
  async stop(projectId: UUID): Promise<void> {
    await this.client.post(`/api/v1/projects/${projectId}/stop`, {});
  }

  /**
   * Pause project execution
   */
  async pause(projectId: UUID): Promise<void> {
    await this.client.post(`/api/v1/projects/${projectId}/pause`, {});
  }

  /**
   * Resume project execution
   */
  async resume(projectId: UUID): Promise<void> {
    await this.client.post(`/api/v1/projects/${projectId}/resume`, {});
  }

  /**
   * Get project logs
   */
  async getLogs(projectId: UUID, limit?: number): Promise<any[]> {
    const params = limit ? { limit } : {};
    return this.client.get(`/api/v1/projects/${projectId}/logs`, params);
  }

  /**
   * Get project metrics
   */
  async getMetrics(projectId: UUID): Promise<any> {
    return this.client.get(`/api/v1/projects/${projectId}/metrics`);
  }

  /**
   * Download project files as an archive
   */
  async download(projectId: UUID): Promise<Blob> {
    const httpClient = this.client.getHttpClient();
    const response = await httpClient.get(`/api/v1/projects/${projectId}/download`, {
      responseType: 'blob',
    });
    return response.data;
  }

  /**
   * Upload files to a project
   */
  async upload(projectId: UUID, file: File | Blob, filename?: string): Promise<void> {
    const formData = new FormData();
    formData.append('file', file, filename);

    const httpClient = this.client.getHttpClient();
    await httpClient.post(`/api/v1/projects/${projectId}/upload`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
  }

  /**
   * Search projects
   */
  async search(
    query: string,
    params: PaginationParams = {}
  ): Promise<PaginatedResponse<Project>> {
    const queryParams = {
      q: query,
      page: params.page ?? 1,
      per_page: params.perPage ?? 20,
    };

    const data = await this.client.get('/api/v1/projects/search', queryParams);
    return this.transformPaginatedResponse<Project>(data);
  }

  /**
   * Get project statistics
   */
  async getStats(): Promise<any> {
    return this.client.get('/api/v1/projects/stats');
  }

  /**
   * Transform API project data to client format
   */
  private transformProject(data: any): Project {
    return {
      id: data.id,
      name: data.name,
      description: data.description,
      techStack: data.tech_stack || [],
      architecture: data.architecture,
      status: data.status,
      createdAt: new Date(data.created_at),
      updatedAt: new Date(data.updated_at),
      metadata: data.metadata || {},
    };
  }

  /**
   * Transform API paginated response to client format
   */
  private transformPaginatedResponse<T>(data: any): PaginatedResponse<T> {
    return {
      data: data.data.map((item: any) => {
        if ('tech_stack' in item) {
          return this.transformProject(item);
        }
        return item;
      }),
      total: data.total,
      page: data.page,
      perPage: data.per_page,
      hasNext: data.has_next,
      hasPrev: data.has_prev,
    };
  }
}

File: sdk/python/src/aion_sdk/projects.py
"""Projects API interface."""

from typing import List, Optional, Dict, Any
from uuid import UUID

from aion_sdk.models import (
    Project,
    ProjectRequest,
    ProjectStatus,
    PaginatedResponse,
    PaginationParams,
    TemplateRequest,
)


class ProjectsAPI:
    """Projects API interface for managing AION projects."""

    def __init__(self, client: "AionClient") -> None:
        """Initialize the projects API."""
        self._client = client

    async def list(
        self,
        page: int = 1,
        per_page: int = 20,
        sort_by: Optional[str] = None,
        sort_order: str = "asc",
    ) -> PaginatedResponse[Project]:
        """List all projects with optional pagination.

        Args:
            page: Page number (default: 1)
            per_page: Items per page (default: 20)
            sort_by: Field to sort by
            sort_order: Sort order ('asc' or 'desc')

        Returns:
            Paginated list of projects
        """
        params = {
            "page": page,
            "per_page": per_page,
            "sort_order": sort_order,
        }
        if sort_by:
            params["sort_by"] = sort_by

        data = await self._client.get("/api/v1/projects", params=params)
        return PaginatedResponse.model_validate(data)

    async def get(self, project_id: UUID) -> Project:
        """Get a specific project by ID.

        Args:
            project_id: The project ID

        Returns:
            The project details
        """
        data = await self._client.get(f"/api/v1/projects/{project_id}")
        return Project.model_validate(data)

    async def create(self, request: ProjectRequest) -> Project:
        """Create a new project.

        Args:
            request: Project creation request

        Returns:
            The created project
        """
        data = await self._client.post("/api/v1/projects", json_data=request.model_dump())
        return Project.model_validate(data)

    async def update(self, project_id: UUID, request: ProjectRequest) -> Project:
        """Update an existing project.

        Args:
            project_id: The project ID
            request: Project update request

        Returns:
            The updated project
        """
        data = await self._client.put(f"/api/v1/projects/{project_id}", json_data=request.model_dump())
        return Project.model_validate(data)

    async def delete(self, project_id: UUID) -> None:
        """Delete a project.

        Args:
            project_id: The project ID
        """
        await self._client.delete(f"/api/v1/projects/{project_id}")

    async def from_template(self, request: TemplateRequest) -> Project:
        """Generate a project from a template.

        Args:
            request: Template generation request

        Returns:
            The generated project
        """
        data = await self._client.post("/api/v1/projects/from-template", json_data=request.model_dump())
        return Project.model_validate(data)

    async def get_status(self, project_id: UUID) -> ProjectStatus:
        """Get project execution status.

        Args:
            project_id: The project ID

        Returns:
            The project status
        """
        data = await self._client.get(f"/api/v1/projects/{project_id}/status")
        return ProjectStatus(data)

    async def start(self, project_id: UUID) -> None:
        """Start project execution.

        Args:
            project_id: The project ID
        """
        await self._client.post(f"/api/v1/projects/{project_id}/start", json_data={})

    async def stop(self, project_id: UUID) -> None:
        """Stop project execution.

        Args:
            project_id: The project ID
        """
        await self._client.post(f"/api/v1/projects/{project_id}/stop", json_data={})

    async def pause(self, project_id: UUID) -> None:
        """Pause project execution.

        Args:
            project_id: The project ID
        """
        await self._client.post(f"/api/v1/projects/{project_id}/pause", json_data={})

    async def resume(self, project_id: UUID) -> None:
        """Resume project execution.

        Args:
            project_id: The project ID
        """
        await self._client.post(f"/api/v1/projects/{project_id}/resume", json_data={})

    async def get_logs(self, project_id: UUID, limit: Optional[int] = None) -> List[Dict[str, Any]]:
        """Get project logs.

        Args:
            project_id: The project ID
            limit: Maximum number of logs to return

        Returns:
            List of log entries
        """
        params = {}
        if limit:
            params["limit"] = limit

        return await self._client.get(f"/api/v1/projects/{project_id}/logs", params=params)

    async def get_metrics(self, project_id: UUID) -> Dict[str, Any]:
        """Get project metrics.

        Args:
            project_id: The project ID

        Returns:
            Project metrics data
        """
        return await self._client.get(f"/api/v1/projects/{project_id}/metrics")

    async def download(self, project_id: UUID) -> bytes:
        """Download project files as an archive.

        Args:
            project_id: The project ID

        Returns:
            Archive file content as bytes
        """
        response = await self._client._client.get(f"/api/v1/projects/{project_id}/download")
        response.raise_for_status()
        return response.content

    async def upload(self, project_id: UUID, file_data: bytes, filename: str) -> None:
        """Upload files to a project.

        Args:
            project_id: The project ID
            file_data: File content as bytes
            filename: Name of the file
        """
        files = {"file": (filename, file_data, "application/octet-stream")}
        response = await self._client._client.post(f"/api/v1/projects/{project_id}/upload", files=files)
        response.raise_for_status()

    async def search(
        self,
        query: str,
        page: int = 1,
        per_page: int = 20,
    ) -> PaginatedResponse[Project]:
        """Search projects.

        Args:
            query: Search query
            page: Page number (default: 1)
            per_page: Items per page (default: 20)

        Returns:
            Paginated search results
        """
        params = {
            "q": query,
            "page": page,
            "per_page": per_page,
        }

        data = await self._client.get("/api/v1/projects/search", params=params)
        return PaginatedResponse.model_validate(data)

    async def get_stats(self) -> Dict[str, Any]:
        """Get project statistics.

        Returns:
            Project statistics
        """
        return await self._client.get("/api/v1/projects/stats")